### ISSUE TITLE
Add Swiss German in multiple dialects

### ISSUE THREAD

**Issue by Roemer** [rhasspy/piper#49]
As a Swiss-German, I would love to have text spoken in it. If possible, even in one of the various dialects.
There is a freely available dataset with 3 hours of high quality speech with transcript in each of the most used swiss-german dialects (8 as of now). They can be found here: https://mtc.ethz.ch/publications/open-source/swiss-dial.html
And here is a sample implementation: https://stt4sg.fhnw.ch/tts
I will try to find out if we could use the dataset for piper and then might need some help building the models.

**Comment by Roemer** [comment]
Ok I got the data set. It is licenses under `Creative Commons Attribution-NonCommercial 4.0 International License` so that should be fine as long as we add attribution somewhere.
No I don't really know how to train the data but would like to do so.
I have a json with the transcribed texts and one where also numerics are written words and for each dialect a folder with 3 hours or more wav files according to the json transcriptions.
Now how can train the langauge model? I am fairly fit in programming so if I need to convert or restructure the data first, that won't be a problem.

**Comment by synesthesiam** [comment]
Hi @Roemer, thanks for the link to the dataset and for giving it a try!

Unfortunately, it looks like [espeak-ng](https://github.com/espeak-ng/espeak-ng) doesn't yet have support for Swiss German (`de-ch`). One option would be to help add this upstream :)

We have a small Swiss German group gathered here for the Home Assistant project: https://github.com/home-assistant/intents/discussions/801
I don't know any other languages besides U.S. English, but I could contribute my linguistics knowledge.

**Comment by dontinelli** [comment]
@synesthesiam, if I understand you correctly, piper relies somehow on eSpeak, so to get de_ch into piper, eSpeak must be updated first.

I'm not yet familiar with neither piper nor eSpeak. Had a short look at the documentation on [adding a language to espeak](https://github.com/espeak-ng/espeak-ng/blob/ae74f733f3334a644b547c39e238f8c1ebfeec63/docs/add_language.md). Isn't straight forward but seems doable. One would just need some time to have a detailed look at the instructions (and for me I would need to get more familiar with the phonetics).

@Roemer, do you already have experience with eSpeak?

**Comment by Roemer** [comment]
@dontinelli No, not yet but it seems doable. What I am unsure about is with our different dialects as they are phonetically quite different and not just "another speaker". Would we need `de-ch-be`, `de-ch-zh`, `de-ch-gr` and such or should one `de-ch` work?

I would be willing to spend some time and/or computing power (got quite some servers in the basement for stuff and free energy) to get that working.

Also there are other self-hosted TTS solutions available like https://github.com/coqui-ai/TTS which seem quite a bit more sophisticated and developed than piper. What is the big benefit of doing all in piper which seems to do nothing more than coqui-ai/TTS?

**Comment by synesthesiam** [comment]
@Roemer You will find similar problems with Coqui-TTS for `de-ch`. Both Piper and Coqui-TTS can train models on raw text, which would get around the immediate need to change espeak-ng. However, this means that numbers, dates, abbreviations, etc. would have to be spelled out before they are sent to the TTS model.

Additionally, I wrote one of Coqui-TTS' underlying phonemizers ([gruut](https://github.com/rhasspy/gruut/)) so I know it doesn't support `de-ch` yet :wink:  

> What is the big benefit of doing all in piper which seems to do nothing more than coqui-ai/TTS?

Piper is optimized for the Raspberry Pi 4, with the core Piper exe + libraries taking 40MB of disk space, most voice models taking less than 75MB, and the ability to run 2x faster than real-time.  Coqui-TTS is indeed more sophisticated, but it is paid for with about 4GB of dependencies just to get started (including needing Python), many voices requiring hundreds of MB, and having large variances in speed depending on the voice/vocoder combo used.

**Comment by dontinelli** [comment]
I'm no expert in TTS but I tend to agree with @synesthesiam. In particular for HA (where my use case is) a lightweight TTS engine is needed (at least in the first instance).

**Comment by Whyki** [comment]
@synesthesiam   Thanks for all the amazing work you have done! I have to say that I have been keenly following multiple of your projects: Larynx, Gruut and now Piper!

I was surprised that you don't use gruut in Piper, and because I have seen you mentioning it here I thought it might be the right moment to ask.  Is there any specific reason to not use gruut? Is that because of it being written in python? Or it is too large?

**Comment by Roemer** [comment]
I started to get into espeak-ng and was actually able to compile it (in a vscode dev-container) and add the needed files for at least one basic Swiss-German (gsw) language and added the numbers 1-12 to try it out. Seems to work good so far but it seems to be a LOT of work to create the `list` and `rules` file.

**Comment by dontinelli** [comment]
> I started to get into espeak-ng and was actually able to compile it (in a vscode dev-container) and add the needed files for at least one basic Swiss-German (gsw) language and added the numbers 1-12 to try it out. Seems to work good so far but it seems to be a LOT of work to create the `list` and `rules` file.
> 
> Edit: Who ever wants to help to add everything needed to espeak-ng can join my repo here: https://github.com/Roemer/espeak-ng/tree/feature/swiss-german
> 
> Also I added an issue in espeak-ng in hope to get some help: [espeak-ng/espeak-ng#1747](https://github.com/espeak-ng/espeak-ng/issues/1747)

It seems you already got further than I did. I once started a draft PR (https://github.com/espeak-ng/espeak-ng/pull/1724) but have not come very far. Without having had a look at your repo, this might be a better basis to continue...
